In [4]:
# 케라스 설치
!pip install keras==2.0.8


In [7]:
!pip install tensorflow==1.13.1

In [8]:
# 버전 확인
!pip list

Package              Version            
-------------------- -------------------
absl-py              0.9.0              
alembic              1.4.2              
astor                0.8.0              
attrs                19.3.0             
backcall             0.1.0              
bleach               3.1.0              
certifi              2019.11.28         
certipy              0.1.3              
colorama             0.4.3              
decorator            4.4.2              
defusedxml           0.6.0              
entrypoints          0.3                
gast                 0.3.3              
grpcio               1.27.2             
h5py                 2.10.0             
importlib-metadata   1.5.0              
ipykernel            5.1.4              
ipython              7.13.0             
ipython-genutils     0.2.0              
jedi                 0.16.0             
Jinja2               2.11.1             
json5                0.9.2              
jsonschema      

In [1]:
import tensorflow as tf
print(tf.__version__)

C:\Users\chlgy\Anaconda3\envs\tf0\lib\site-packages\tensorflow\python\framework\dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\Users\chlgy\Anaconda3\envs\tf0\lib\site-packages\tensorflow\python\framework\dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\Users\chlgy\Anaconda3\envs\tf0\lib\site-packages\tensorflow\python\framework\dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\Users\chlgy\Anaconda3\envs\tf0\lib\site-packages\tensorflow\python\framework

1.13.1


In [2]:
import keras
keras.__version__

Using TensorFlow backend.


'2.0.8'

**keras** 에는 이미 mnist 데이터가 포함되어 있어, 다음과 같이 데이터를 가져올 수 있다. 

In [3]:
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [4]:
train_images.shape
#shape를 확인한 결과 28 * 28 픽셀의 60000개의 training images가 있다.

(60000, 28, 28)

In [5]:
len(train_labels)
# train_labels는 0-9사이의 숫자 배열로 60000개가 있다.

60000

In [6]:
train_labels
# train_labels의 배열은 다음과 같다.

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [7]:
test_images.shape
# test_images의 shape을 확인한 결과 28 * 28 픽셀의 10000개 데이터를 가지고 있다.

(10000, 28, 28)

In [8]:
len(test_labels)
# test_labels 역시 0-9 사이의 숫자이며 10000개가 있다. 

10000

In [9]:
test_labels
# test_labels의 배열은 다음과 같다.

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

In [10]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

Instructions for updating:
Colocations handled automatically by placer.


딥러닝 모델은 데이터 처리를 위한 체와 같은 것으로, 점점 더 정교한 데이터 필터인 "레이어"로 이루어져 있다. 여기서는 두 개의 Dense layers로 구성되어 있다. keras에서는 첫번째 레이어에서 데이터 형태 input_shape를 지정해 준다.   

첫번째 Dense layer(완전 연결층)은 512개의 노드(뉴런)과 활성화 함수로 relu를 사용한다.  

마지막의 소프트맥스 층의 경우는 10개의 확률 점수가 들어있는 배열을 반환한다.

In [11]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

Instructions for updating:
keep_dims is deprecated, use keepdims instead


신경망이 훈련 준비를 마치기 위해서는 컴파일 단계에 포함될 3가지가 더 필요하다.  

- loss function: 네트워크가 훈련 데이터에서 얼마나 일을 잘 하고 있는지를 측정할 수 있다. (여기서는 범주형 크로스 엔트로피=softmax loss를 사용했다.)    


- Optimizer: 데이터와 손실함수에 기초하여 스스로 업데이트하는 메커니즘이다.(여기서는 RMSProp을 사용했다.)



- 훈련과 테스트 과정을 모니터링할 지표 (여기서는 정확도를 사용했다)

In [12]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

훈련에 앞서 네트워크가 기대하는 형태로 데이터를 재편성하고, 모든 값이 [0, 1] 간격으로 되도록 스케일링을 하여 데이터를 미리 처리한다.   

이전에, 우리의 훈련 이미지는 [0, 255] 간격의 값과 함께 uint8 형식의 배열(600, 28, 28)로 저장되었다. 우리는 이것을 0과 1 사이의 값으로 float32 배열(60000, 28 * 28)로 변환한다.

In [13]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

우리는 label들을 카테고리로 인코딩 시켜줘야 한다. 

In [14]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)
#입력데이터(train_images)와 출력데이터(train_labels)를 인자로 넣고, 작업 단위 128로 5번 훈련시킨다.

Instructions for updating:
Use tf.cast instead.
Epoch 1/5
60000/60000 [==============================] - 7s - loss: 0.2544 - acc: 0.9278     
Epoch 2/5
60000/60000 [==============================] - 7s - loss: 0.1021 - acc: 0.9685     
Epoch 3/5
60000/60000 [==============================] - 7s - loss: 0.0675 - acc: 0.9797     
Epoch 4/5
60000/60000 [==============================] - 6s - loss: 0.0487 - acc: 0.9852     
Epoch 5/5
60000/60000 [==============================] - 7s - loss: 0.0375 - acc: 0.9887     - ETA: 0s - loss: 0.0371 - 


훈련 중에 두 가지 quantity가 표시되는데, 그것은 훈련 데이터에 대한 네트워크의 "손실"과 훈련 데이터에 대한 네트워크의 정확성이다.

 

우리는 훈련 데이터에 대해 0.9892(즉, 98.9%)의 정확도에 빠르게 도달한다. 

In [18]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
#테스트 세트에서 모델의 성능을 비교

 9760/10000 [============================>.] - ETA: 0s

In [19]:
print('test_acc:', test_acc)

test_acc: 0.9771


우리의 시험 세트 정확도는 97.8%로 나타났는데, 이는 훈련 세트 정확도보다 상당히 낮은 것이다. 훈련 정확도와 시험 정확도 사이의 이러한 차이는 훈련세트에 '과대적합' 되었기 때문이다.